src/index.ts

```diff
@@ -99,6 +100,68 @@ export const jsonSchemaTransform = createJsonSchemaTransform({
   skipList: defaultSkipList,
 });
 
+export const createJsonSchemaTransformObject =
+  ({ schemas: zodSchemas }: { schemas: Record<string, z.ZodTypeAny> }) =>
+  (
+    input:
+      | { swaggerObject: Partial<OpenAPIV2.Document> }
+      | { openapiObject: Partial<OpenAPIV3.Document | OpenAPIV3_1.Document> },
+  ) => {
+    if ('swaggerObject' in input) {
+      console.warn('This package currently does not support component references for Swagger 2.0');
+      return input.swaggerObject;
+    }
+
+    const schemas: Record<string, OpenAPIV3.SchemaObject | OpenAPIV3_1.SchemaObject> = {};
+    for (const key in zodSchemas)
+      schemas[key] = zodToJsonSchema(zodSchemas[key], zodToJsonSchemaOptions);
+
+    const document = {
+      ...input.openapiObject,
+      components: {
+        ...input.openapiObject.components,
+        schemas: {
+          ...input.openapiObject.components?.schemas,
+          ...schemas,
+        },
+      },
+    };
+
+    const componentMapVK = new Map<string, string>();
+    Object.entries(schemas).forEach(([key, value]) =>
+      componentMapVK.set(JSON.stringify(value), key),
+    );
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    function componentReplacer(this: any, key: string, value: any) {
+      if (typeof value !== 'object') return value;
+
+      // Check if the parent is the schemas object, if so, return the value as is
+      if (this === document.components.schemas) return value;
+
+      const stringifiedValue = JSON.stringify(value);
+      if (componentMapVK.has(stringifiedValue))
+        return { $ref: `#/components/schemas/${componentMapVK.get(stringifiedValue)}` };
+
+      if (value.nullable === true) {
+        const nonNullableValue = { ...value };
+        delete nonNullableValue.nullable;
+        const stringifiedNonNullableValue = JSON.stringify(nonNullableValue);
+        if (componentMapVK.has(stringifiedNonNullableValue))
+          return {
+            anyOf: [
+              { $ref: `#/components/schemas/${componentMapVK.get(stringifiedNonNullableValue)}` },
+            ],
+            nullable: true,
+          };
+      }
+
+      return value;
+    }
+
+    return JSON.parse(JSON.stringify(document, componentReplacer));
```

Comment:
what is the purpose of this? deep cloning?

Reply:
It checks if the stringified object exists in the "componentMapVK" of the provided schemas, by checking if the stringified values are equal to eachother. It's a quick way of checking if objects match each other on all levels.

Reply:
can you extract it into an internal function called areDeepEqual to make it explicit?

Reply:
can you add the comment explaining that we are doing the deep equality check then/